src/module/canvas/ruler.ts

```diff
@@ -199,38 +196,52 @@ class RulerPF2e<TToken extends TokenPF2e | null = TokenPF2e | null> extends Rule
             return super._highlightMeasurementSegment(segment);
         }
 
-        this.#highlighted.clear();
-        if (canvas.grid.isHexagonal) {
-            this.#highlightHexSegment(segment);
-        } else {
-            this.#highlightSegment(segment);
-        }
-    }
+        // There is no point of highlighting a segment on a gridless grid
+        if (!canvas.grid.isGridless) {
+            // This is used to store the offset from the token's center to the center of the grid we want to work from.
+            // On a square grid this is not needed, if a point is in the center of four squares then the offest of that point represents the square in the top-left.
+            // However, on a hexagonal grid this is not the same if the point is in the center of three hexes. Depending on the circumstances it can think the point is in any of the three hexes due to floating point math.
+            // So what we are going to do is ensure we are always using the square/hex that is to the top-left of the center point if the token's center is inbetween squares/hexes.
+            let centerOffset = { x: 0, y: 0 };
+
+            // Check to see if the token's width is even, this mess the center point is going to be touching multiple square/hexes
+            if (Math.max(token.document.width, 1) % 2 !== 1) {
+                // Check to see if the grid is square, if so we want to offset the center by half the grid
+                if (canvas.grid.isSquare) {
+                    centerOffset = { x: canvas.grid.size / 2, y: canvas.grid.size / 2 };
+                }
 
-    #highlightSegment(segment: RulerMeasurementSegment): void {
-        // Keep track of grid spaces set to be highlighed in order to skip repeated highlighting
-        for (const offset of canvas.grid.getDirectPath([segment.ray.A, segment.ray.B])) {
-            for (const stomp of this.#footprint) {
-                const newOffset = { i: offset.i + stomp.i, j: offset.j + stomp.j };
-                const packed = (newOffset.i << 16) + newOffset.j;
-                if (!this.#highlighted.has(packed)) {
-                    this.#highlighted.add(packed);
-                    const point = canvas.grid.getTopLeftPoint(newOffset);
-                    canvas.interface.grid.highlightPosition(this.name, { ...point, color: this.color });
+                // Check to see if the grid is hexagonal
+                if (canvas.grid.isHexagonal) {
+                    if ((canvas.grid as HexagonalGrid).columns) {
```

Comment:
Is this assertion necessary? I vaguely recall putting in place a discriminated union.